asm/vm/asm_vm

```diff
@@ -518,62 +536,203 @@ EOF
+    false
+  fi
+}
+
+is_membership_crd_installed() {
+  if ! kubectl api-resources --api-group=hub.gke.io | grep -q memberships; then
+    false
+  fi
+}
+
+validate_workloadgroup() {
+  info "Checking for WorkloadGroup ${WORKLOAD_NAME} in namespace ${WORKLOAD_NAMESPACE}..."
   local RESULT; RESULT=""
+  RESULT="$(retry 2 kubectl get workloadgroups.networking.istio.io "${WORKLOAD_NAME}" \
+    -n "${WORKLOAD_NAMESPACE}" || true)"
+  if [[ -z "${RESULT}" ]]; then
+    { read -r -d '' MSG; fatal "${MSG}"; } <<EOF || true
+Unable to find WorkloadGroup ${WORKLOAD_NAME} in namespace ${WORKLOAD_NAMESPACE}.
+Please verify the name and namespace of the WorkloadGroup and try again.
+EOF
+  fi
+
+  parse_workloadgroup
+  validate_workload_asm_revision
+  validate_workloadgroup_identity_provider
+  validate_workload_service_account
+}
+
+parse_workloadgroup() {
+  retrieve_workload_asm_revision
+  retrieve_workload_labels
+  retrieve_workload_service_account
+}
+
+validate_workloadgroup_identity_provider() {
+  info "Verifying the identity provider of the WorkloadGroup..."
+  WORKLOAD_IDENTITY_PROVIDER="$(retry 2 kubectl get workloadgroups.networking.istio.io \
+    "${WORKLOAD_NAME}" -n "${WORKLOAD_NAMESPACE}" -ojson \
+    | jq -r '.spec.metadata.annotations["'"${IDENTITY_PROVIDER_ANNOTATION_KEY}"'"]')"
+  if [[ -z "${WORKLOAD_IDENTITY_PROVIDER}" ]]; then
+    { read -r -d '' MSG; fatal "${MSG}"; } <<EOF || true
+Unable to find the identity provider for WorkloadGroup ${WORKLOAD_NAME} in
+namespace ${WORKLOAD_NAMESPACE}.
+Please make sure an IdentityProvider is specified in the WorkloadGroup.
+EOF
+  fi
 
-  RESULT="$(kubectl get service istio-dns -n kube-system || true)"
+  local RESULT; RESULT=""
+  RESULT="$(retry 2 kubectl get identityproviders.security.cloud.google.com \
+    "${WORKLOAD_IDENTITY_PROVIDER}" || true)"
   if [[ -z "${RESULT}" ]]; then
-    fatal "istio-dns service not found on the cluster. Please follow the user
-guide and create a DNS service for VMs."
+    { read -r -d '' MSG; fatal "${MSG}"; } <<EOF || true
+Unable to find identity provider ${WORKLOAD_IDENTITY_PROVIDER} in the cluster.
+To see the available identity providers in the cluster, run:
+  kubectl get identityproviders.security.cloud.google.com
+EOF
   fi
+}
 
-  if [[ "${CREATE_GCE_INSTANCE_TEMPLATE}" -eq 0 ]]; then
-    return 0
+retrieve_workload_asm_revision() {
+  info "Retrieving the ASM revision for the WorkloadGroup..."
+  WORKLOAD_ASM_REVISION="$(retry 2 kubectl get namespace "${WORKLOAD_NAMESPACE}" -ojson \
+    | jq -r '.metadata.labels["'${ASM_REVISION_LABEL_KEY}'"]')"
+  if [[ -z "${WORKLOAD_ASM_REVISION}" ]]; then
+    { read -r -d '' MSG; fatal "${MSG}"; } <<EOF || true
+Unable to find an ASM revision for the namespace of the WorkloadGroup. To apply
+a revision label on the namespace, replace <REVISION> with an available revision
+name and run:
+  kubectl label namespace "${WORKLOAD_NAMESPACE}" istio-injection- \
```

Comment:
These double quotes will be printed literally since you're in a heredoc.

Reply:
Removed.